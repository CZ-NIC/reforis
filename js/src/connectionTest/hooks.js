/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useCallback, useEffect, useMemo, useState } from "react";

import API_URLs from "common/API";
import { useAPIPost, useWSForisModule } from "foris";

const TESTS_TYPES = {
    wan: ["ipv6", "ipv6_gateway", "ipv4", "ipv4_gateway"],
    IPv6Disabled: ["ipv4", "ipv4_gateway"],
    dns: ["dns", "dnssec"],
    overview: ["ipv6", "ipv6_gateway", "ipv4", "ipv4_gateway", "dns", "dnssec"],
};

const ENDPOINTS = {
    wan: API_URLs.connectionTest,
    dns: API_URLs.dnsTest,
    // dnsTest is also called for the overview type because it returns both wan and dns results
    overview: API_URLs.dnsTest,
};

export const TEST_STATES = {
    NOT_RUNNING: 0,
    RUNNING: 1,
    FINISHED: 2,
};

export default function useConnectionTest(ws, type) {
    const initialResults = useMemo(
        () =>
            TESTS_TYPES[type].reduce((tests, test) => {
                tests[test] = null;
                return tests;
            }, {}),
        [type]
    );

    const [state, setState] = useState(TEST_STATES.NOT_RUNNING);
    const [id, setId] = useState(null);
    const [results, setResults] = useState(initialResults);

    const updateResults = useCallback(
        (data) => {
            if (data && data.test_id === id) {
                setResults((prevTestResults) => ({
                    ...prevTestResults,
                    ...filterResults(data.data, type),
                }));
                if (data.passed) setState(TEST_STATES.FINISHED);
            }
        },
        [id, type]
    );

    const wsModule = "wan";
    const [wsData] = useWSForisModule(ws, wsModule, "connection_test");
    useEffect(() => {
        updateResults(wsData);
    }, [wsData, id, type, updateResults]);
    const [wsFinishedData] = useWSForisModule(
        ws,
        wsModule,
        "connection_test_finished"
    );
    useEffect(() => {
        updateResults(wsFinishedData);
    }, [wsFinishedData, id, type, updateResults]);

    const [triggerTestData, triggerTest] = useAPIPost(ENDPOINTS[type]);
    useEffect(() => {
        if (triggerTestData.data) {
            setState(TEST_STATES.RUNNING);
            setResults(initialResults);
            setId(triggerTestData.data.test_id);
        }
    }, [initialResults, triggerTestData]);

    return [state, results, triggerTest];
}

function filterResults(results, type) {
    console.log("results", results);
    return Object.keys(results)
        .filter((test) => TESTS_TYPES[type].indexOf(test) !== -1)
        .reduce((res, test) => {
            res[test] = results[test];
            return res;
        }, {});
}
