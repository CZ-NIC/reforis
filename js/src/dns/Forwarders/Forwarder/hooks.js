/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect } from "react";
import update from "immutability-helper";
import {
    ALERT_TYPES,
    API_STATE,
    useAlert,
    useAPIPost,
    useAPIPut,
    useForm,
} from "foris";
import API_URLs from "common/API";
import validator from "./validator";

const EMPTY_FORWARDER = {
    name: "",
    description: "",
    ipaddresses: {
        ipv4: [""],
        ipv6: [""],
    },
    tls_type: "no",
};

export default function useForwarderForm(forwarder, saveForwarderCallback) {
    const [formState, setFormValue, initForm] = useForm(validator);
    const [postState, post] = useAPIPost(API_URLs.dnsForwarders);
    const [putState, put] = useAPIPut(
        `${API_URLs.dnsForwarders}/${(forwarder || {}).name}`
    );

    const [setAlert, dismissAlert] = useAlert();

    useEffect(() => {
        if (postState.state === API_STATE.SUCCESS) {
            saveForwarderCallback();
            initForm(EMPTY_FORWARDER);
            setAlert(_("Forwarder saved successfully."), ALERT_TYPES.SUCCESS);
        } else if (postState.state === API_STATE.ERROR) {
            setAlert(_("Can't save forwarder."));
        }
    }, [setAlert, postState, initForm, saveForwarderCallback]);

    useEffect(() => {
        if (putState.state === API_STATE.SUCCESS) {
            saveForwarderCallback();
            setAlert(_("Forwarder added successfully."), ALERT_TYPES.SUCCESS);
        } else if (putState.state === API_STATE.ERROR) {
            setAlert(_("Can't add new forwarder."));
        }
    }, [putState, saveForwarderCallback, setAlert]);

    useEffect(() => {
        if (forwarder) {
            initForm(forwarder);
        } else {
            initForm(EMPTY_FORWARDER);
        }
    }, [forwarder, initForm]);

    function saveForwarder() {
        const data = prepDataToSubmit(formState.data);
        dismissAlert();
        if (forwarder) {
            put({ data });
        } else {
            post({ data });
        }
    }

    return [formState, setFormValue, postState, saveForwarder];
}

function prepDataToSubmit(forwarder) {
    const tlsUnsetRules = {
        no: ["tls_hostname", "tls_pin"],
        hostname: ["tls_pin"],
        pin: ["tls_hostname"],
    };
    const fieldsToUnset = ["name", "editable"].concat(
        tlsUnsetRules[forwarder.tls_type]
    );
    const ipAddresses = filterEmptyIPAddresses(forwarder.ipaddresses);
    return update(forwarder, {
        $unset: fieldsToUnset,
        ipaddresses: { $set: ipAddresses },
    });
}

function filterEmptyIPAddresses(ipaddresses) {
    return {
        ipv4: ipaddresses.ipv4.filter((ipaddress) => !!ipaddress),
        ipv6: ipaddresses.ipv6.filter((ipaddress) => !!ipaddress),
    };
}
