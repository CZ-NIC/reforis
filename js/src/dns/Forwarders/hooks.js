/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect, useState } from "react";
import update from "immutability-helper";
import { useAPIGet, useWSForisModule, API_STATE } from "foris";

import API_URLs from "common/API";

const MODULE = "dns";

export default function useForwardersList(ws) {
    const [forwarders, setForwarders] = useState([]);

    const [forwardersListState, getForwardersList] = useAPIGet(
        API_URLs.dnsForwarders
    );
    useEffect(() => {
        getForwardersList();
    }, [getForwardersList]);

    useEffect(() => {
        if (forwardersListState.state === API_STATE.SUCCESS) {
            setForwarders(forwardersListState.data.forwarders || []);
        }
    }, [forwardersListState]);

    useForwardersWS(ws, setForwarders);

    return [sortForwarders(forwarders), forwardersListState.state];
}

function sortForwarders(forwarders) {
    return forwarders.sort((a, b) => {
        if (a.editable === b.editable) {
            return a.name.localeCompare(b.name);
        }
        return a.editable ? 1 : -1;
    });
}

function useForwardersWS(ws, setForwarders) {
    function useForwarderWSAction(action, func) {
        const [wsForwarderData] = useWSForisModule(ws, MODULE, action);
        useEffect(() => {
            if (wsForwarderData)
                setForwarders((forwarders) =>
                    func(forwarders, wsForwarderData)
                );
        }, [func, wsForwarderData]);
    }

    useForwarderWSAction("add_forwarder", addForwarder);
    useForwarderWSAction("set_forwarder", setForwarder);
    useForwarderWSAction("del_forwarder", deleteForwarder);
}

function addForwarder(forwarders, forwarder) {
    return update(forwarders, { $push: [{ editable: true, ...forwarder }] });
}

function setForwarder(forwarders, forwarder) {
    const forwarderIndex = forwarders.findIndex(
        (fwd) => fwd.name === forwarder.name
    );
    return update(forwarders, {
        $splice: [[forwarderIndex, 1, { editable: true, ...forwarder }]],
    });
}

function deleteForwarder(forwarders, forwarder) {
    const forwarderIndex = forwarders.findIndex(
        (fwd) => fwd.name === forwarder.name
    );
    return update(forwarders, { $splice: [[forwarderIndex, 1]] });
}
