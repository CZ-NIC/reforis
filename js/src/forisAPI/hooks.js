/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useState} from 'react';

import ForisAPI from "./forisAPI";

export function useAPIGetData(endpoint) {
    const [isReady, setIsReady] = useState(false);

    function getData(callback) {
        setIsReady(false);
        ForisAPI[endpoint].get().then(data => {
            callback(data);
            setIsReady(true);
        });
    }

    return [getData, isReady]
}

export function useAPIPostData(endpoint) {
    function postData(data) {
        ForisAPI[endpoint].post(data).then(
            data => {
                console.log(data) //TODO: remove
            }
        );
    }

    return postData
}
