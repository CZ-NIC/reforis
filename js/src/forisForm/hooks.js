/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useReducer} from 'react';
import update from 'immutability-helper';

export const FORM_STATES = {
    READY: 1,
    UPDATE: 2,
    NETWORK_RESTART: 3,
    LOAD: 4,
};

export const FORM_ACTIONS = {
    setData: 1,
    setState: 2,
};

function formReducer(state, action) {
    console.log(action); //TODO: remove
    switch (action.type) {
        case FORM_ACTIONS.setData: {
            return {
                data: action.data,
                errors: action.errors,
                state: FORM_STATES.READY,
            };
        }
        case FORM_ACTIONS.setState:
            return {
                ...state,
                state: action.state
            };
        default:
            return state;
    }
}

export function useForm(prepData, validator) {
    const [state, dispatch] = useReducer(formReducer,
        {
            data: {},
            errors: {},
            state: FORM_STATES.LOAD
        }
    );

    function getChangedValue(target) {
        let value = target.value;
        if (target.type === 'checkbox')
            value = target.checked;
        else if (target.type === 'number')
            value = parseInt(value);
        return value
    }

    function setFormValue(updateRule) {
        return event => {
            const value = getChangedValue(event.target);
            const newData = update(state.data, updateRule(value));
            dispatch({
                type: FORM_ACTIONS.setData,
                data: newData,
                errors: validator(newData),
            });
        };
    }

    function setFormData(data) {
        const newData = prepData(data);
        dispatch({
            type: FORM_ACTIONS.setData,
            data: newData,
            errors: validator(newData),
        });
    }

    function setFormState(state) {
        dispatch({
            type: FORM_ACTIONS.setState,
            state: state,
        });
    }

    function isDisabled() {
        return state.state !== FORM_STATES.READY;
    }

    return [
        state.data,
        state.errors,
        setFormData,

        state.state,
        setFormState,

        isDisabled(),
        setFormValue,
        dispatch,
    ]
}