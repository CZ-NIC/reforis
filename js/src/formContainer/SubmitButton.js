/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import Button from 'common/bootstrap/Button';

export const STATES = {
    READY: 1,
    SAVING: 2,
    LOAD: 3,
};

SubmitButton.propTypes = {
    disabled: propTypes.bool,
    state: propTypes.oneOf(Object.keys(STATES).map(key => STATES[key])),
    remindsToNWRestart: propTypes.number,
};

export default function SubmitButton({disabled, state, remindsToNWRestart, ...props}) {
    const disableSubmitButton = disabled || state !== STATES.READY;
    const loadingSubmitButton = state !== STATES.READY;

    let labelSubmitButton;
    switch (state) {
        case STATES.SAVING:
            labelSubmitButton = _('Updating');
            break;
        case STATES.LOAD:
            labelSubmitButton = _('Load settings');
            break;
        default:
            labelSubmitButton = _('Save');
    }

    return <Button
        loading={loadingSubmitButton}
        disabled={disableSubmitButton}
        forisFormSize

        {...props}
    >
        {labelSubmitButton}
    </Button>;
}
