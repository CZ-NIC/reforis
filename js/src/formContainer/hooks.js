/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {useCallback, useEffect, useReducer} from 'react';
import update from 'immutability-helper';

import {useAPIGet} from '../common/APIhooks';
import {useWSForisModule} from '../common/WebSocketsHooks';

export function useForm(validator, prepData) {
    const [state, dispatch] = useReducer(formReducer, {
        data: null,
        initialData: null,
        errors: {},
    });

    useEffect(() => {
        bindUnsavedChangesAlert(state.data, state.initialData);
    }, [state.data, state.initialData]);

    const onFormReload = useCallback(data => {
        dispatch({
            type: FORM_ACTIONS.resetData,
            data: data,
            prepData: prepData,
            validator: validator,
        });
    }, [prepData, validator]);

    const onFormChangeHandler = useCallback(updateRule =>
        event => {
            dispatch({
                type: FORM_ACTIONS.updateValue,
                value: getChangedValue(event.target),
                updateRule: updateRule,
                validator: validator,
            })
        }, [validator]);

    return [
        state,
        onFormChangeHandler,
        onFormReload,
    ]
}

const FORM_ACTIONS = {
    updateValue: 1,
    resetData: 2,
};

function formReducer(state, action) {
    switch (action.type) {
        case FORM_ACTIONS.updateValue:
            const newData = update(state.data, action.updateRule(action.value));
            const errors = action.validator(newData);
            return {
                ...state,
                data: newData,
                errors: errors
            };
        case FORM_ACTIONS.resetData:
            if (!action.data)
                return {...state, initialData: state.data};
            const prepData = action.prepData ? action.prepData(action.data) : action.data;
            return {
                data: prepData,
                initialData: prepData,
                errors: action.data ? action.validator(prepData) : undefined,
            };
        default:
            throw new Error();
    }
}

function getChangedValue(target) {
    let value = target.value;
    if (target.type === 'checkbox') {
        value = target.checked;
    } else if (target.type === 'number') {
        const parsedValue = parseInt(value);
        value = isNaN(parsedValue) ? value : parsedValue;
    }
    return value
}

function bindUnsavedChangesAlert(currentData, initialData) {
    window.onbeforeunload = e => {
        if (JSON.stringify(currentData) === JSON.stringify(initialData)) {
            return;
        }
        const message = _('Changes you made may not be saved.'),
            event = e || window.event;
        if (event) {
            event.returnValue = message;
        }
        return message;
    };
}

export function useForisModule(ws, config) {
    const [APIGetState, get] = useAPIGet(config.endpoint);
    const [WSData] = useWSForisModule(ws, config.wsModule, config.wsAction);

    useEffect(() => {
        get();
    }, [WSData, get]);

    return [APIGetState];
}
