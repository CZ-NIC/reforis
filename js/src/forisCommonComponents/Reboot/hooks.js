/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {useEffect, useState} from 'react';

import {fetchWithTimeout} from '../../forisAPI/utils';
import {ForisPort, ForisProtocol} from '../../constants';

export const REBOOT_STATES = {
    NOT_TRIGGERED: 0,
    TRIGGERED: 1,
    IN_PROCESS: 2,
    DONE: 3,
};

const HEALTH_CHECK_URL = '/api/health-check';
const HEALTH_CHECK_TIMEOUT = 5000;

const sleep = (milliseconds) => {
    return new Promise(resolve => setTimeout(resolve, milliseconds))
};

export default function useRebootHandling(ws) {
    const [state, setState] = useState(REBOOT_STATES.NOT_TRIGGERED);
    const [ips, setIPs] = useState([]);
    const [remainsSec, setRemainsSec] = useState(null);

    useEffect(() => {
        const wsModule = 'maintain';
        ws.subscribe(wsModule)
            .bind(wsModule, 'reboot', msg => {
                setRemainsSec(msg.data.remains / 1000);
                if (msg.data.remains === 0) {
                    setState(REBOOT_STATES.IN_PROCESS);
                } else {
                    setIPs([...new Set(msg.data.ips)]);
                    setState(REBOOT_STATES.TRIGGERED);
                }
            })
    }, []);

    useEffect(() => {
        if (state === REBOOT_STATES.IN_PROCESS) {
            ws.ws.close();
            waitForDeath();
        } else if (state === REBOOT_STATES.DONE) {
            tryReconnect();
        }
    }, [state]);

    function waitForDeath() {
        fetchWithTimeout(HEALTH_CHECK_URL, undefined, HEALTH_CHECK_TIMEOUT)
            .then(async () => {
                await sleep(1000);
                waitForDeath();
            })
            .catch((e) => {
                setState(REBOOT_STATES.DONE);
            })
    }

    function tryReconnect() {
        ips.forEach(async ip => {
            //TODO: PORT is not needed on prod.
            const url = `${ForisProtocol}://${ip}:${ForisPort}`;
            waitForLife(url);
        });
    }

    function waitForLife(url) {
        fetchWithTimeout(`${url}${HEALTH_CHECK_URL}`, {}, HEALTH_CHECK_TIMEOUT)
            .then(() => {
                window.location.replace(url);
            })
            .catch((e) => {
                waitForLife(url);
            });
    }

    return [state, remainsSec];
}
