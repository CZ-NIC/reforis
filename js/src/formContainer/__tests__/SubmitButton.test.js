/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {render} from 'react-testing-library'

import SubmitButton from '../SubmitButton';
import {FORM_STATES} from '../hooks';

describe('<SubmitButton/>', () => {
    it('Render ready', () => {
        const {container} = render(<SubmitButton state={FORM_STATES.READY}/>);
        expect(container).toMatchSnapshot();
    });
    it('Render load', () => {
        const {container} = render(<SubmitButton state={FORM_STATES.LOAD}/>);
        expect(container).toMatchSnapshot();
    });
    it('Render update', () => {
        const {container} = render(<SubmitButton state={FORM_STATES.UPDATE}/>);
        expect(container).toMatchSnapshot();
    });
    it('Render update', () => {
        const {container} = render(<SubmitButton state={FORM_STATES.NETWORK_RESTART} remindsToNWRestart={1234}/>);
        expect(container).toMatchSnapshot();
    });
});
