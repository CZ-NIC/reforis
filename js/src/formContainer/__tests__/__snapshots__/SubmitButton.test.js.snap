// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubmitButton/> Render load 1`] = `
<div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    disabled=""
  >
    <span
      aria-hidden="true"
      class="spinner-border spinner-border-sm"
      role="status"
    />
     
     
     
    Load settings
  </button>
</div>
`;

exports[`<SubmitButton/> Render ready 1`] = `
<div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
  >
     
     
    Save
  </button>
</div>
`;

exports[`<SubmitButton/> Render saving 1`] = `
<div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    disabled=""
  >
    <span
      aria-hidden="true"
      class="spinner-border spinner-border-sm"
      role="status"
    />
     
     
     
    Updating
  </button>
</div>
`;
